Code Challenge : Membantu siswa dalam mengasa kemampuan berfikir kritis menyelesaikan kasus dalam bentuk pemrograman

# Code Challenge

1. Tulis program Python sederhana yang meminta pengguna untuk memasukkan dua angka, kemudian menghitung dan mencetak hasil penjumlahan kedua angka tersebut.

In [ ]:
input_1 = input("Masukkan angka pertama : ")
input_2 = input("Masukkan angka kedua : ")
hasil = int(input_1) +int( input_2)
print("Hasil dari penjumlahan", input_1 , "dan" , input_2 ,"adalah" , hasil)

Masukkan angka pertama : 2
Masukkan angka kedua : 5
Hasil dari penjumlahan 2 dan 5 adalah 7


2. Buatlah program Python yang meminta pengguna untuk memasukkan jari-jari lingkaran, dan kemudian menghitung dan mencetak luas lingkaran tersebut. Gunakan rumus luas lingkaran: Luas = π * r^2, dengan π (pi) bernilai 3.14.

In [ ]:
r = input('Masukan jari-jari lingkaran : ')
luas=3.14*int(r)**2
print("Luas lingkaran =",luas)

Masukan jari-jari lingkaran : 4
Luas lingkaran = 50.24


3. Buatlah program Python yang meminta pengguna untuk memasukkan sebuah bilangan bulat positif, lalu menghitung dan mencetak faktorial dari bilangan tersebut. Faktorial dari bilangan bulat n (dilambangkan dengan n!) adalah hasil perkalian semua bilangan bulat positif dari 1 hingga n.

  Example : <br>
Masukkan bilangan bulat positif: 5
Faktorial dari 5 adalah: 120


In [ ]:
def menghitung_faktorial(n):
    if n < 0:
        return "Faktorial tidak didefinisikan untuk bilangan negatif"
    elif n == 0 or n == 1:
        return 1
    else:
        faktorial = 1
        for i in range(1, n + 1):
            faktorial *= i
        return faktorial
input_n = int(input('Masukan bilangan positif : '))
hasil_faktorial = menghitung_faktorial(input_n)
print("Faktorial dari", input_n , 'adalah' , hasil_faktorial)



Masukan bilangan positif : 5
Faktorial dari 5 adalah 120


4. Tulis program Python yang meminta pengguna memasukkan sebuah bilangan bulat positif, dan kemudian menentukan apakah bilangan tersebut adalah bilangan prima atau bukan. Sebuah bilangan prima hanya memiliki dua faktor positif, yaitu 1 dan dirinya sendir

  Example :
Masukkan bilangan bulat positif: 7
7 adalah bilangan prima.

  Masukkan bilangan bulat positif: 12
12 bukan bilangan prima.

In [ ]:
def cek_prima(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Meminta pengguna untuk memasukkan bilangan bulat positif
bilangan = int(input("Masukkan bilangan bulat positif: "))

# Menentukan apakah bilangan tersebut adalah bilangan prima
if cek_prima(bilangan):
    print(f"{bilangan} adalah bilangan prima.")
else:
    print(f"{bilangan} bukan bilangan prima.")

Masukkan bilangan bulat positif: 8
8 bukan bilangan prima.


In [ ]:
def cek_prima(n):
    if n < 0:
        return "Faktorial tidak didefinisikan untuk bilangan negatif"
    elif n == 0 or n == 1:
        return 1
    else:
        faktorial = 1
        for i in range(1, n + 1):
            faktorial *= i
        return faktorial
input_n = int(input('Masukan bilangan positif : '))
hasil_l = menghitung_faktorial(input_n)
print("Faktorial dari", input_n , 'adalah' , hasil_faktorial)

5. Buatlah program Python yang meminta pengguna untuk memasukkan sebuah kata atau frase, dan kemudian menentukan apakah kata atau frase tersebut merupakan palindrom (bisa dibaca sama dari depan maupun dari belakang).

  Masukkan sebuah kata atau frase: radar
radar adalah palindrom.

  Masukkan sebuah kata atau frase: hello
hello bukan palindrom.

In [ ]:
def cek_palindrom(kata):
  cleaned=kata.replace(' ','').lower()
  return cleaned==cleaned[::-1]
masukan=input("Masukan kata: ")
if cek_palindrom(masukan):
  print(f'{masukan} adalah palindrom')
else:
  print(f'{masukan} bukan palindrom')

Masukan kata: bab
bab adalah palindrom


6. Buatlah sebuah program yang mengeluarkan output berupa profile seperti nama, kota asal, band favorit dan makan favorit.

  Example :
Nama : Zen
Domisili : Kendari
Band Favorit : JKT48
Makanan Favorit : Nasi Goreng

In [ ]:
input_nama=input('Masukan nama : ')
input_domisili=input('Masukan domisili : ')
input_band=input('Masukan band favorit : ')
input_makanan=input('Masukan makanan favorit : ')

print(f'Nama : {input_nama} Domisili : {input_domisili} Band Favorit : {input_band} Makanan Favorit : {input_makanan} ')

Masukan nama : ubay
Masukan domisili : sleman
Masukan band favorit : flow
Masukan makanan favorit : indomie
Nama : ubay Domisili : sleman Band Favorit : flow Makanan Favorit : indomie 


7. Buatlah sebuah program yang dapat menghitung luas total segitiga

In [ ]:
input_p = int(input('Masukan Panjang Segitiga:'))
input_t = int(input('Masukan Tinggi Segitiga:'))

luas=(input_p*input_t)/2

print(f'Luas Segitiga = {luas}')

Masukan Panjang Segitiga:3
Masukan Tinggi Segitiga:8
Luas Segitiga = 12.0



8. Pak maman ingin menukarkan uangnya sejumlah RP. 32.892.700 kedalam pecahan uang Rp. 100.000 Rp 50.000 Rp. 10.000 Rp 5.000 Rp 1000 Rp 500 dan Rp 100 ke pak budi, Buatlah pemrograman python yang bisa membantu mempercepat pak Budi dalam menentukan berapa buah/lembar uang untuk masih masih pecahan tersebut

In [ ]:
def tukar_pecahan(uang):
    pecahan = [100000, 50000, 10000, 5000, 1000, 500, 100]
    hasil = {}

    for nominal in pecahan:
        jumlah = uang // nominal
        uang %= nominal
        hasil[nominal] = jumlah
    return hasil
tukar=32892700
hasil_tukar=tukar_pecahan(tukar)

print(f"Jumlah uang yang ditukar: Rp. {tukar:,}")
print("Rincian pecahan:")
for nominal in hasil_tukar:
    jumlah = hasil_tukar[nominal]
    print(f"Rp. {nominal:,}   : {jumlah} lembar")


Jumlah uang yang ditukar: Rp. 32,892,700
Rincian pecahan:
Rp. 100,000   : 328 lembar
Rp. 50,000   : 1 lembar
Rp. 10,000   : 4 lembar
Rp. 5,000   : 0 lembar
Rp. 1,000   : 2 lembar
Rp. 500   : 1 lembar
Rp. 100   : 2 lembar


9. Buatlah program python yang berguna untuk mengkonversi dari celcius menjadi fahrenheit

In [ ]:
celcius=input("Masukan suhu dalam celcius : ")
fahrenheit=(int(celcius)/5)*9+32
print(f'{celcius} C = {fahrenheit} F')

Masukan suhu dalam celcius : 0
0 C = 32.0 F


10. Supermarket X adalah satu-satunya supermarket di kota Solo terkenal karena kejujurannya. Berapapun total belanja pelanggan, selalu dibulatkan ke bawah ke kelipatan Rp 1000 rupiah. Sebagai contoh misalkan total belanja pelanggan adalah Rp 120.580, maka akan dibulatkan ke Rp 120.000 atau misalkan totalnya Rp 99.999 maja akan dibulatkan kebawah menjadi Rp 99.000. Karena kebaikan supermarket inilah,m semakin lama semakin banyak pelanggannya. Namun, karena kasirnya masih menggunakan sistem manual, maka semakin lama kasir mengalami kerepotan dalam menangani transaksi perharinya. Buatlah sebuah pemrograman Python untuk membantu kasir dalam menentukan total belanja sesuai ketentuan yang berlaku di supermarket tersebut. Input program adalah total belanja pelanggan, sedangkan outputnya adalah total belanja setelah pembulatan ke bawah ke kelipatan Rp 1000

In [ ]:
def pembulatan_ke_ribuan(angka):
    return int(angka / 1000) * 1000

input_harga = input("Masukkan beberapa harga, pisahkan dengan spasi: ")

harga_list = [int(harga) for harga in input_harga.split()]

total = sum(harga_list)
hasil_pembulatan = pembulatan_ke_ribuan(total)

print(f"Total belanja: Rp {total:,}")
print(f"Total yang dibayar: Rp {hasil_pembulatan:,}")

Masukkan beberapa harga, pisahkan dengan spasi: 6655 7782 5245 52555
Total belanja: Rp 72,237
Total yang dibayar: Rp 72,000


11. Hitunglah Upah Kerja (Kotor) dari seluruh karyawan Perusahaan B jika diketahui upah per jam adalah Rp 30000

  Tampilkan output sebagai berikut

  Contoh:
Upah Kerja (Kotor) Doni = Rp 380000
Keterangan :
  - Upah per jam = 30000
  - Upah Kerja (Kotor) = Jam Kerja x Upah per Jam
Upah Kerja (Kotor) Joko = Rp 1320000
Upah Kerja (Kotor) Alex = Rp 1440000
Upah Kerja (Kotor) Chandra = Rp 1500000
Upah Kerja (Kotor) Hartono = Rp 1260000
Upah Kerja (Kotor) Samsul = Rp 1350000

In [ ]:
def upah_kotor(jam, upah_perjam):
  return jam*upah_perjam
karyawan = {
    'Doni' : 10,
    'Joko' : 15,
    'Alex' : 8,
    'Chandra' : 13,
    'Hartono' : 11,
    'Samsul' : 9
}
upah_perjam=30000
for nama, jam in karyawan.items():
  upah=upah_kotor(jam,upah_perjam)
  print(f"Upah kerja (kotor) {nama} = Rp {upah:,}")

Upah kerja (kotor) Doni = Rp 300,000
Upah kerja (kotor) Joko = Rp 450,000
Upah kerja (kotor) Alex = Rp 240,000
Upah kerja (kotor) Chandra = Rp 390,000
Upah kerja (kotor) Hartono = Rp 330,000
Upah kerja (kotor) Samsul = Rp 270,000


12. Hitunglah Upah Lembur dari seluruh karyawan Perusahaan B

  Tampilkan output sebagai berikut: contoh

  Contoh:
Upah Lembur Doni = Rp 23000
Keterangan :
  - Upah lembur per jam = 35000
  - Upah Lembur = Jam Lembur x Upah lembur per jam
Upah Lembur Joko = Rp 420000
Upah Lembur Alex = Rp 385000
Upah Lembur Chandra = Rp 350000
Upah Lembur Hartono = Rp 525000
Upah Lembur Samsul = Rp 420000

In [ ]:
def upah_lembur(jam, upah_perjam):
  return jam*upah_perjam
karyawan = {
    'Doni' : 4,
    'Joko' : 5,
    'Alex' : 3.8,
    'Chandra' : 1.3,
    'Hartono' : 1,
    'Samsul' : 4
}
upah_perjam=35000
for nama, jam in karyawan.items():
  upah=upah_lembur(jam,upah_perjam)
  print(f"Upah kerja (lembur) {nama} = Rp {upah:,}")

Upah kerja (lembur) Doni = Rp 140,000
Upah kerja (lembur) Joko = Rp 175,000
Upah kerja (lembur) Alex = Rp 133,000.0
Upah kerja (lembur) Chandra = Rp 45,500.0
Upah kerja (lembur) Hartono = Rp 35,000
Upah kerja (lembur) Samsul = Rp 140,000


13. Kita berbelanja bahan makanan di supermarket. Berikut merupakan harga bahan makanan yang dibeli:

  ayam adalah 15000
bumbu 10000
beras 50000.
Total belanja nya adalah 160000.
Cobalah masukkan angka tersebut ke dalam variabel, kemudian print menggunakan f-string untuk semua variabel!

In [ ]:
belanja = {"ayam":15000,"bumbu":10000,"beras":50000,"Total belanja":160000,}
for nama,harga in belanja.items():
  print(f"{nama} : Rp {harga:,}")

ayam : Rp 15,000
bumbu : Rp 10,000
beras : Rp 50,000
Total belanja : Rp 160,000


Rima merupakan seorang anak yang sangat hemat. Namun, suatu ketika dirinya merasa boros dan akhirnya Rima pun me-list pengeluarannya minggu ini dan didapat hasil sebagai berikut:

Baju: 200.000
Tumblr: 75.000
Tas: 150.000
Berapakah total biaya yang telah dikeluarkan oleh Rima dalam minggu ini?

In [ ]:
list_khilaf = {"Baju":200000,"Tumblr":75000,"Tas":15000,}
total_khilaf = sum(list_khilaf.values())

for nama,harga in list_khilaf.items():
  print(f"{nama} : Rp {harga:,}")
print(f"Total belanja: Rp {total_khilaf:,}")


Baju : Rp 200,000
Tumblr : Rp 75,000
Tas : Rp 15,000
Total belanja: Rp 290,000


14. James Hoffman pergi ke PaperCup Coffee milik Andi, setelah mencicipi kopi buatan toko tersebut James termotivasi untuk membuat kopi sendiri ketika di rumah nanti. James pergi ke Supermarket untuk membeli keperluan membuat kopi, berikut adalah barang yang dibeli dan harganya:

Biji Kopi yang dibeli adalah "Ethiopian Coffee Beans" dengan harga Rp 115.000
Alat yang dibutuhkan untuk menyeduh kopi sendiri James membeli alat "V60" dengan harga Rp 95.500
Cobalah masukkan nilai - nilai tersebut ke dalam variable dan buatlah code yang sesuai dengan output!

output : Biji Kopi yang dibeli oleh James Hoffman adalah Ethiopian Coffee Beans dengan harga 115000
Alat untuk menyeduh kopi yang dibeli oleh James Hoffman adalah V60 dengan harga 95500

In [ ]:
list_belanja = {"Ethiopian Coffee Beans":115000, "V60":95500}
biji_kopi = "Ethiopian Coffee Beans"
alat_seduh = "V60"

print(f"Biji Kopi yang dibeli oleh James Hoffman adalah {biji_kopi} dengan harga Rp {list_belanja[biji_kopi]:,}")
print(f"Alat untuk menyeduh kopi yang dibeli oleh James Hoffman adalah {alat_seduh} dengan harga Rp {list_belanja[alat_seduh]:,}")

Biji Kopi yang dibeli oleh James Hoffman adalah Ethiopian Coffee Beans dengan harga Rp 115,000
Alat untuk menyeduh kopi yang dibeli oleh James Hoffman adalah V60 dengan harga Rp 95,500


15. Pak Hartono agak kesal karena harga BBM naik dari Rp 7650 per liter menjadi Rp 10000 per liter. Jika sebelumnya beliau membeli Rp 20000 sudah full tank, berapa liter yang harus dibayar Pak Hartono jika mengisi full tank dengan harga yang baru ?

In [ ]:
x =10000*(20000/7650)
hasil=int(x / 1000) * 1000
print (f"Rp {hasil:,}")

Rp 26,000


16. Putra bekerja di sebuah perusahaan swasta. Beliau mendapatkan gaji tahunan sebesar Rp 160 juta dan dibayarkan setiap dua minggu. Jika dalam satu tahun gaji tersebut dibayarkan sebanyak 26 kali, berapakah gaji yang di dapatkan Putra selama dua minggu ?

In [ ]:
gaji_pertahun = 160000000
total_pembayaran_pertahun = 26

gaji_2minggu= gaji_pertahun / total_pembayaran_pertahun

print(f"Gaji yang diterima Putra setiap dua minggu adalah Rp {gaji_2minggu:,.2f}")

Gaji yang diterima Putra setiap dua minggu adalah Rp 6,153,846.15


17. Anda memiliki perusahaan percetakan. Pelanggan anda meminta anda untuk mencetak tulisan "HAPPI BIRTHDAY". Namun ternyata di dalam file tulisan yang dikirim, anda menyadari bahwa ada tulisan typo yaitu I di dalam kata HAPPI yang seharusnya huruf Y. Pelanggan meminta anda untuk mengganti dengan yang benar dan mengubah hurufnya menjadi besar di awal kata saja. Coba revisi tulisan HAPPI BIRTHDAY sesuai permintaan pelanggan anda!

In [2]:
def revisi(kata):
  cleaned=kata.replace(' ','').lower()


18. Terdapat kode pos NZ12348, dimana 2 karakter awal adalah kode negara. Coba ambil kode negara dari string tersebut!

In [1]:
kode_pos="NZ12348"
kode_negara=kode_pos(0,2)
print(kode_negara)

TypeError: 'str' object is not callable

19. Pengguna aplikasi e-commerce mendapatkan voucher belanja jika memenuhi salah satu ketentuan yaitu belanja 7 hari berturut-turut atau belanja lebih dari 1 juta, Hilmi tidak belanja 7 hari berturut-turut namun ia telah melakukan transaksi 1 juta, apakah Hilmi mendapatkan voucher belanja? (True/False)

In [ ]:
or


20. Pak Joko membandingkan harga stock produk barang kecantikan pada kedua supplier yang berbeda.
Supplier pertama menawarkan 10 kardus isi produk kecantikan Nivae dengan harga 1.500.000 rupiah
Sedangkan supplier kedua menawarkan 12 kardus isi produk kecantikan Garneir dengan harga 1.700.000 rupiah.
Dengan kedua harga tersebut Pak Joko memprioritaskan untuk menekan budgetnya dengan membeli stock produk yang paling murah yaitu dengan cara mempertimbangkan harga per kardusnya.
Buatlah operasi matematika perbandingan stock produk barang kecantikan apakah harga produk Nivae lebih besar dari harga produk Garneir dalam Python!

21. ----------------------------------------------------------------------
Start Up ditempat Haris bekerja dengan timnya sedang melakukan pengujian testing aplikasinya pada fitur dalam mendapatkan voucher belanja untuk user  

| Jenis Voucher        |Nama variabel    |Tipe Data|
|----------------------|-----------------|---------|
|Voucher Gratis Ongkir |get_free_shipping|Boolean  |
|Voucher Cashback      |get_cashback     |Boolean  |
|Voucher Potongan Harga|get_discount     |Boolean  |
|Voucher Pengguna Baru |get_newuser      |Boolean  |

Terdapat beberapa user jenis yang mendapatkan voucher:
1. Pengguna baru mendapatkan voucher pengguna baru saja
2. Pengguna lama yang mendapatkan voucher gratis ongkir saja
3. pengguna lama yang mendapatkan voucher potongan harga saja
4. pengguna lama yang membeli barang mahal mendapatkan voucher gratis ongkir,cashback
5. pengguna lama yang membeli barang mahal mendapatkan voucher gratis ongkir, voucher cashback dan potongan harga

semua pengguna harus menggunakan semua variabel yang berada pada tabel <br>

21A. Buatlah variabel yang sesuai dengan pengguna baru mendapatkan voucher pengguna baru saja <br>
21B. Buatlah variabel pengguna lama yang mendapatkan voucher gratis ongkir saja <br>
21C. Buatlah variabel pengguna lama yang mendapatkan voucher potongan harga saja <br>
21D. Buatlah variabel pengguna lama yang membeli barang mahal mendapatkan voucher gratis ongkir,cashback <br>
21E. Buatlah variabel pengguna lama yang membeli barang mahal mendapatkan voucher gratis ongkir, voucher cashback dan potongan harga

22. Anda makan di sebuah restoran. Anda ingin memesan menu sesuai dengan budget anda.

Jika uang anda lebih dari sama dengan 100.000 maka anda memakan steak
Jika uang anda lebih dari sama dengan 50.000 maka anda memakan spaghetti
Jika uang anda kurang dari 50.000 maka anda memakan nasi goreng
Budget uang anda sebesar 60.000 maka lengkapilah program branching dibawah ini

```
uang = ...
... ... ... ...:
    print("Makan Steak")
... ... ... ...:
    ...
else:
    ...
```

23. Jika Universitas yang dipilih adalah ITB, maka akan masuk FMIPA
Jika Universitas yang dipilih adalah UI, maka akan masuk Statistik
Jika tidak memasuki Universitas selain kondisi di atas, maka Yuni akan mencari Universitas yang lain

```
universitas = "UI"
... ...:
    ...
... ...:
    print("Jurusan statistik")
else:
    print("Cari Universitas yang lain, semangat!!!")
```

24. Lengkapi program branching dibawah ini,

Jika nilai lebih dari sama dengan 75 maka akan mendapatkan nilai A
Jika nilai ada di 50-75 maka akan mendapatkan nilai B
Jika nilai di bawah 50 maka akan mendapatkan nilai C.

```
nilai = 80
... ... >= ...:
    ...
... ... ... 50:
    print("nilai B")
...:
    ...
```

25. Sebuah Toko Florist memberikan diskon 20% kepada pelanggan yang telah melakukan 3 transaksi sebelumnya
Akan memberi diskon 30% untuk pelanggan yang telah melakukan 7 kali transaksi sebelumnya
Gamal ingin membeli bouquet bunga dan sekarang merupakan transaksi yang ke 7 di Toko Florist tersebut
Buat program Branching di bawah ini berdasarkan case diatas!

26. Dokter merawat seorang pasien yang sedang sakit, dokter memiliki beberapa kondisi diagnosa yaitu

Jika pasien flu tanpa gejala sakit tenggorokan bahwa pasien sakit flu biasa
Jika pasien flu ditambah dengan sakit tenggorokan maka pasien sakit radang tenggorokan
Jika pasien flu ditambah sakit tenggorokan ditambah mual maka pasien sakit tonsilitis
Ternyata terdapat pasien yang mengidap radang tenggorokan. Ubahlah variabel kondisi_1 dan kondisi_2 agar mengeluarkan diagnosa "pasien sakit radang tenggorokan"

27. Pada Supermarket A customer akan mendapatkan promo menarik jika:

Jika customer berbelanja di atas Rp 350.000 akan mendapatkan diskon 5%
Jika customer berbelanja di atas Rp 350.000 dan barang yang terakhir dibeli adalah "Sate Cap Windbreaker" akan mendapatkan diskon 15%
Jika customer berbelanja tidak di atas Rp 350.000 tidak mendapatkan diskon